archive/archivist/src/main/scala/uk/ac/wellcome/platform/archive/archivist/flow/S3UploadFlow.scala

```diff
@@ -7,20 +7,25 @@ import akka.stream._
 import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
 import akka.util.ByteString
 import com.amazonaws.services.s3.AmazonS3
-import com.amazonaws.services.s3.model._
+import com.amazonaws.services.s3.model.{ObjectMetadata => S3ObjectMetadata, _}
 import grizzled.slf4j.Logging
+import uk.ac.wellcome.platform.archive.archivist.models.storage.ObjectMetadata
 import uk.ac.wellcome.storage.ObjectLocation
 
 import scala.annotation.tailrec
 import scala.collection.JavaConverters._
 import scala.util.{Failure, Success, Try}
 
 object S3UploadFlow extends Logging {
-  def apply(uploadLocation: ObjectLocation)(implicit s3Client: AmazonS3) =
-    new S3UploadFlow(uploadLocation)(s3Client)
+  def apply(uploadLocation: ObjectLocation,
+            maybeUploadMetadata: Option[ObjectMetadata] = None)(
```

Comment:
minor: I wonder if it would be cleaner to make this a parameter that defaults to an empty map? Something like:

```suggestion
            uploadMetadata: ObjectMetadata = ObjectMetadata(Map.empty))(
```

The `toS3ObjectMetadata` method should work fine (you just get an empty metadata blob), and it reduces the number of Options you have to pass around.

Actually, what are the scenarios when we wouldn't want metadata? (Aren't we always annotating uploads with hashes?)

Reply:
The scenario where we don't want metadata is uploading the tagManifest does not include a checksum and hence no metadata.  We could default to an empty map.

Reply:
I also think the SUploadFlow is fairly generic and we should eventually move into scala-storage, so I wouldn't enforce metadata

Reply:
yup, thanks @alicefuzier.